File: backend/project-service/src/main/java/com/solarease/service/DemandService.java
package com.solarease.service;

import com.solarease.dto.DemandCreateRequest;
import com.solarease.dto.DemandDTO;
import com.solarease.dto.PublicDemandCreateRequest;
import com.solarease.dto.ProjectRequest;
import com.solarease.dto.ProjectResponse;
import com.solarease.entity.Client;
import com.solarease.entity.DemandEntity;
import com.solarease.enums.DemandPriority;
import com.solarease.enums.DemandSource;
import com.solarease.enums.DemandStatus;
import com.solarease.exception.ResourceNotFoundException;
import com.solarease.repository.ClientRepository;
import com.solarease.repository.DemandRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Locale;

@Service
@RequiredArgsConstructor
@Slf4j
public class DemandService {

    private final DemandRepository demandRepository;
    private final ProjectService projectService;
    private final ClientRepository clientRepository;
    private final NotificationWebSocketService notificationService;
    private final InvitationService invitationService;
    private final IdentityServiceClient identityServiceClient;

    // ── CLIENT: submit a new demand ──────────────────────────────────────────

    @Transactional
    public DemandDTO createDemand(
            String clientUserId,
            String clientEmail,
            String clientFirstName,
            String clientLastName,
            DemandCreateRequest req) {

        DemandEntity entity = DemandEntity.builder()
                .clientUserId(clientUserId)
                .clientEmail(clientEmail)
                .clientFirstName(clientFirstName)
                .clientLastName(clientLastName)
                .status(DemandStatus.NOUVELLE)
                .source(DemandSource.CLIENT)
                .priority(DemandPriority.NORMALE)
                .name(req.getName())
                .description(req.getDescription())
                .location(req.getLocation())
                .latitude(req.getLatitude())
                .longitude(req.getLongitude())
                .peakPower(req.getPeakPower())
                .availableArea(req.getAvailableArea())
                .inclination(req.getInclination())
                .orientation(req.getOrientation())
                .budget(req.getBudget())
                .build();

        DemandEntity saved = demandRepository.save(entity);
        DemandDTO dto = toDto(saved);
        notificationService.notifyAdminsOnNewDemand(dto);
        return dto;
    }

    @Transactional
    public DemandDTO createPublicDemand(PublicDemandCreateRequest req) {
        String fullName = req.getFullName() == null ? "" : req.getFullName().trim();
        String[] names = splitFullName(fullName);

        String subjectLabel = mapSubject(req.getSubject());
        String demandName = "Demande " + subjectLabel + " - " + (names[0].isBlank() ? "Prospect" : names[0]);
        String email = req.getEmail().trim().toLowerCase(Locale.ROOT);

        StringBuilder description = new StringBuilder();
        if (req.getMessage() != null && !req.getMessage().isBlank()) {
            description.append(req.getMessage().trim());
        }
        if (req.getPhone() != null && !req.getPhone().isBlank()) {
            if (!description.isEmpty()) {
                description.append("\n\n");
            }
            description.append("Téléphone: ").append(req.getPhone().trim());
        }

        DemandEntity entity = DemandEntity.builder()
                .clientUserId("PUBLIC:" + email)
                .clientEmail(email)
                .clientFirstName(names[0])
                .clientLastName(names[1])
                .clientPhone(req.getPhone())
                .status(DemandStatus.NOUVELLE)
                .source(DemandSource.PUBLIC)
                .priority(DemandPriority.NORMALE)
                .name(demandName)
                .description(description.toString())
                .build();

        DemandEntity saved = demandRepository.save(entity);
        DemandDTO dto = toDto(saved);
        notificationService.notifyAdminsOnNewDemand(dto);
        return dto;
    }

    // ── CLIENT: list own demands ─────────────────────────────────────────────

    public Page<DemandDTO> getMyDemands(String clientUserId, Pageable pageable) {
        return demandRepository.findByClientUserId(clientUserId, pageable).map(this::toDto);
    }

    // ── ADMIN: list / search demands ─────────────────────────────────────────

    public Page<DemandDTO> getAllDemands(Pageable pageable) {
        return demandRepository.findAll(pageable).map(this::toDto);
    }

    public Page<DemandDTO> getDemandsByStatus(DemandStatus status, Pageable pageable) {
        return demandRepository.findByStatus(status, pageable).map(this::toDto);
    }

    /** Inbox-grade search supporting optional status, source and free-text query. */
    public Page<DemandDTO> searchDemands(DemandStatus status,
                                         DemandSource source,
                                         String q,
                                         Pageable pageable) {
        String normalisedQ = (q == null || q.isBlank()) ? null : q.trim();
        return demandRepository.search(status, source, normalisedQ, pageable).map(this::toDto);
    }

    public DemandDTO getDemandById(Long demandId) {
        return demandRepository.findById(demandId)
                .map(this::toDto)
                .orElseThrow(() -> new ResourceNotFoundException("Demand not found: " + demandId));
    }

    // ── ADMIN: change status ──────────────────────────────────────────────────

    @Transactional
    public DemandDTO updateStatus(Long demandId, DemandStatus newStatus, String adminNote, String rejectionReason) {
        DemandEntity demand = demandRepository.findById(demandId)
                .orElseThrow(() -> new ResourceNotFoundException("Demand not found: " + demandId));
        DemandStatus previousStatus = demand.getStatus();
        demand.setStatus(newStatus);
        if (adminNote != null) demand.setAdminNote(adminNote);
        if (rejectionReason != null) demand.setRejectionReason(rejectionReason);
        DemandEntity saved = demandRepository.save(demand);
        DemandDTO dto = toDto(saved);
        if (previousStatus != newStatus) {
            notificationService.notifyDemandStatusChange(dto);
            notificationService.notifyAdminDashboardRefresh("DEMAND_STATUS_CHANGED");
        }
        return dto;
    }

    // ── ADMIN: assign / unassign ──────────────────────────────────────────────

    @Transactional
    public DemandDTO assignDemand(Long demandId, String adminId) {
        DemandEntity demand = demandRepository.findById(demandId)
                .orElseThrow(() -> new ResourceNotFoundException("Demand not found: " + demandId));
        demand.setAssignedAdminId(adminId == null || adminId.isBlank() ? null : adminId);
        return toDto(demandRepository.save(demand));
    }

    @Transactional
    public DemandDTO updatePriority(Long demandId, DemandPriority priority) {
        DemandEntity demand = demandRepository.findById(demandId)
                .orElseThrow(() -> new ResourceNotFoundException("Demand not found: " + demandId));
        demand.setPriority(priority == null ? DemandPriority.NORMALE : priority);
        return toDto(demandRepository.save(demand));
    }

    // ── ADMIN: convert demand to project ─────────────────────────────────────

    /**
     * Full conversion: admin supplies the target clientId (project-service client record).
     */
    @Transactional
    public ProjectResponse convertToProject(
            Long demandId,
            Long clientId,
            String adminId,
            String adminEmail,
            Double latitude,
            Double longitude) {
        DemandEntity demand = demandRepository.findById(demandId)
                .orElseThrow(() -> new ResourceNotFoundException("Demand not found: " + demandId));

        if (demand.getProjectId() != null) {
            throw new IllegalStateException("Demand " + demandId + " has already been converted to project " + demand.getProjectId());
        }

        Double effectiveLatitude = latitude != null ? latitude : demand.getLatitude();
        Double effectiveLongitude = longitude != null ? longitude : demand.getLongitude();

        ProjectRequest projectReq = ProjectRequest.builder()
                .name(demand.getName())
                .description(demand.getDescription())
                .location(demand.getLocation())
            .latitude(effectiveLatitude)
            .longitude(effectiveLongitude)
                .peakPower(demand.getPeakPower())
                .availableArea(demand.getAvailableArea())
                .inclination(demand.getInclination())
                .orientation(demand.getOrientation())
                .budget(demand.getBudget())
                .clientId(clientId)
                .build();

        ProjectResponse project = projectService.createProject(adminId, adminEmail, adminId, adminEmail, projectReq);

        if (latitude != null) {
            demand.setLatitude(latitude);
        }
        if (longitude != null) {
            demand.setLongitude(longitude);
        }
        demand.setProjectId(project.getId());
        demand.setStatus(DemandStatus.VALIDEE);
        demandRepository.save(demand);

        DemandDTO dto = toDto(demand);
        notificationService.notifyDemandStatusChange(dto);

        log.info("Demand {} converted to project {}", demandId, project.getId());
        return project;
    }

    /**
     * Convenience flow for PUBLIC prospects: create an items_client row from the
     * demand fields (or reuse an existing client with the same email), then
     * convert. Returns the newly-created project.
     */
    @Transactional
    public ProjectResponse promotePublicAndConvert(
            Long demandId,
            String adminId,
            String adminEmail,
            Double latitude,
            Double longitude) {

        DemandEntity demand = demandRepository.findById(demandId)
                .orElseThrow(() -> new ResourceNotFoundException("Demand not found: " + demandId));

        String email = demand.getClientEmail();
        if (email == null || email.isBlank()) {
            throw new IllegalStateException("Cannot promote a public demand without a client email");
        }
        String normalizedEmail = email.trim().toLowerCase(Locale.ROOT);

        java.util.Optional<Client> existingClient = clientRepository.findByEmail(normalizedEmail);
        if (existingClient.isEmpty()) {
            existingClient = clientRepository.findByEmail(email);
        }

        Long clientId = existingClient
                .map(Client::getId)
                .orElseGet(() -> {
                    Client created = Client.builder()
                            .firstName(safe(demand.getClientFirstName(), "Prospect"))
                            .lastName(safe(demand.getClientLastName(), "Public"))
                            .email(normalizedEmail)
                            .phoneNumber(demand.getClientPhone())
                            .address(demand.getLocation())
                            .installerId(adminId)
                            .build();
                    Client saved = clientRepository.save(created);
                    log.info("Promoted public demand {} to new client {}", demandId, saved.getId());
                    return saved.getId();
                });

        Client client = clientRepository.findById(clientId)
                .orElseThrow(() -> new ResourceNotFoundException("Client not found: " + clientId));

        ProjectResponse project = convertToProject(demandId, clientId, adminId, adminEmail, latitude, longitude);

        if (client.getUserId() == null || client.getUserId().isBlank()) {
            DemandEntity refreshed = demandRepository.findById(demandId)
                    .orElseThrow(() -> new ResourceNotFoundException("Demand not found: " + demandId));
            invitationService.sendInvitationForDemand(refreshed, client, project.getId());
            project.setInvitationSent(true);
            boolean hasAccount = identityServiceClient.emailHasPortalAccount(client.getEmail());
            // #region agent log
            com.solarease.debug.DebugTrace.log("H1-H4", "DemandService.promotePublicAndConvert",
                    "after invitation send",
                    java.util.Map.of(
                            "clientId", clientId,
                            "clientUserId", "null",
                            "identityHasAccount", hasAccount,
                            "projectId", project.getId(),
                            "emailDomain", email.contains("@") ? email.substring(email.indexOf('@')) : "unknown"
                    ));
            // #endregion
            project.setInvitationMessage(hasAccount
                    ? "Lien de connexion envoyé à " + client.getEmail() + " (compte existant)"
                    : "Invitation inscription envoyée à " + client.getEmail());
        } else {
            project.setInvitationSent(false);
            project.setInvitationMessage("Le client possède déjà un compte portail");
        }

        return project;
    }

    @Transactional
    public void resendInvitation(Long demandId, Long projectId, String message) {
        DemandEntity demand = demandRepository.findById(demandId)
                .orElseThrow(() -> new ResourceNotFoundException("Demand not found: " + demandId));
        Long effectiveProjectId = projectId != null ? projectId : demand.getProjectId();
        if (effectiveProjectId == null) {
            throw new IllegalStateException("Aucun projet associé à cette demande");
        }
        invitationService.resendInvitationForDemand(demand, effectiveProjectId, message);
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private DemandDTO toDto(DemandEntity e) {
        Boolean clientHasAccount = null;
        if (e.getClientEmail() != null && !e.getClientEmail().isBlank()) {
            clientHasAccount = clientRepository.findByEmail(e.getClientEmail().trim().toLowerCase(Locale.ROOT))
                    .map(c -> c.getUserId() != null && !c.getUserId().isBlank())
                    .orElse(false);
        }

        java.util.Optional<java.time.LocalDateTime> sentAt = invitationService.latestInvitationSentAt(e.getId());

        return DemandDTO.builder()
                .id(e.getId())
                .clientUserId(e.getClientUserId())
                .clientEmail(e.getClientEmail())
                .clientFirstName(e.getClientFirstName())
                .clientLastName(e.getClientLastName())
                .clientPhone(e.getClientPhone())
                .status(e.getStatus())
                .source(e.getSource())
                .priority(e.getPriority())
                .assignedAdminId(e.getAssignedAdminId())
                .name(e.getName())
                .description(e.getDescription())
                .location(e.getLocation())
                .latitude(e.getLatitude())
                .longitude(e.getLongitude())
                .peakPower(e.getPeakPower())
                .availableArea(e.getAvailableArea())
                .inclination(e.getInclination())
                .orientation(e.getOrientation())
                .budget(e.getBudget())
                .rejectionReason(e.getRejectionReason())
                .adminNote(e.getAdminNote())
                .projectId(e.getProjectId())
                .clientHasAccount(clientHasAccount)
                .invitationSent(sentAt.isPresent())
                .invitationSentAt(sentAt.orElse(null))
                .createdAt(e.getCreatedAt())
                .updatedAt(e.getUpdatedAt())
                .build();
    }

    private String[] splitFullName(String fullName) {
        if (fullName.isBlank()) {
            return new String[]{"", ""};
        }
        String[] parts = fullName.split("\\s+");
        if (parts.length == 1) {
            return new String[]{parts[0], ""};
        }
        String firstName = parts[0];
        String lastName = String.join(" ", java.util.Arrays.copyOfRange(parts, 1, parts.length));
        return new String[]{firstName, lastName};
    }

    private String mapSubject(String subject) {
        if (subject == null) return "devis";
        return switch (subject.trim().toLowerCase(Locale.ROOT)) {
            case "devis" -> "devis";
            case "info" -> "information";
            case "sav" -> "SAV";
            default -> "autre";
        };
    }

    private static String safe(String value, String fallback) {
        return (value == null || value.isBlank()) ? fallback : value.trim();
    }
}

File: SolarEase Web App Design (2)/src/app/services/demandService.ts
import api from "./api";

export type DemandStatus = "NOUVELLE" | "A_COMPLETER" | "VALIDEE" | "REJETEE";
export type DemandSource = "PUBLIC" | "CLIENT";
export type DemandPriority = "HAUTE" | "NORMALE" | "BASSE";

export interface Demand {
  id: number;
  clientUserId: string;
  clientEmail: string;
  clientFirstName: string;
  clientLastName: string;
  clientPhone?: string;
  status: DemandStatus;
  source: DemandSource;
  priority: DemandPriority;
  assignedAdminId?: string | null;
  name: string;
  description?: string;
  location?: string;
  latitude?: number;
  longitude?: number;
  peakPower?: number;
  availableArea?: number;
  inclination?: number;
  orientation?: number;
  budget?: number;
  rejectionReason?: string;
  adminNote?: string;
  projectId?: number;
  clientHasAccount?: boolean;
  invitationSent?: boolean;
  invitationSentAt?: string;
  createdAt: string;
  updatedAt: string;
  /** Alias kept for compatibility with older code paths. */
  phone?: string;
}

export interface DemandPage {
  content: Demand[];
  totalElements: number;
  totalPages: number;
  number: number;
  size: number;
}

export interface DemandCreateRequest {
  name: string;
  description?: string;
  location?: string;
  latitude?: number;
  longitude?: number;
  peakPower?: number;
  availableArea?: number;
  inclination?: number;
  orientation?: number;
  budget?: number;
}

export interface PublicDemandCreateRequest {
  fullName: string;
  email: string;
  phone?: string;
  subject: string;
  message: string;
}

export interface DemandSearchParams {
  page?: number;
  size?: number;
  status?: DemandStatus;
  source?: DemandSource;
  q?: string;
}

export const STATUS_LABELS: Record<DemandStatus, { label: string; color: string; bg: string; ring: string }> = {
  NOUVELLE:    { label: "Nouvelle",    color: "text-blue-700",   bg: "bg-blue-100",   ring: "ring-blue-200" },
  A_COMPLETER: { label: "À compléter", color: "text-orange-700", bg: "bg-orange-100", ring: "ring-orange-200" },
  VALIDEE:     { label: "Validée",     color: "text-green-700",  bg: "bg-green-100",  ring: "ring-green-200" },
  REJETEE:     { label: "Rejetée",     color: "text-red-700",    bg: "bg-red-100",    ring: "ring-red-200" },
};

export const SOURCE_LABELS: Record<DemandSource, { label: string; short: string; color: string; bg: string }> = {
  PUBLIC: { label: "Formulaire public", short: "Public",  color: "text-purple-700", bg: "bg-purple-100" },
  CLIENT: { label: "Portail client",    short: "Portail", color: "text-slate-700",  bg: "bg-slate-100" },
};

export const PRIORITY_LABELS: Record<DemandPriority, { label: string; color: string; bg: string }> = {
  HAUTE:   { label: "Haute",   color: "text-red-700",    bg: "bg-red-50" },
  NORMALE: { label: "Normale", color: "text-slate-600",  bg: "bg-slate-50" },
  BASSE:   { label: "Basse",   color: "text-emerald-700", bg: "bg-emerald-50" },
};

const demandService = {
  /** CLIENT: submit a new request */
  createDemand: async (data: DemandCreateRequest): Promise<Demand> => {
    const res = await api.post<Demand>("/demands", data);
    return res.data;
  },

  /** PUBLIC: submit a demand from the public contact page */
  createPublicDemand: async (data: PublicDemandCreateRequest): Promise<Demand> => {
    const res = await api.post<Demand>("/demands/public", data);
    return res.data;
  },

  /** CLIENT: list my demands */
  getMyDemands: async (page = 0, size = 10): Promise<DemandPage> => {
    const res = await api.get<DemandPage>(`/demands/my?page=${page}&size=${size}`);
    return res.data;
  },

  /** ADMIN: search/list demands with optional filters */
  getAllDemands: async (
    pageOrParams: number | DemandSearchParams = 0,
    size = 25,
    status?: DemandStatus
  ): Promise<DemandPage> => {
    // Back-compat: original signature was getAllDemands(page, size, status)
    const params: DemandSearchParams =
      typeof pageOrParams === "number"
        ? { page: pageOrParams, size, status }
        : pageOrParams;

    const search = new URLSearchParams();
    search.set("page", String(params.page ?? 0));
    search.set("size", String(params.size ?? 25));
    if (params.status) search.set("status", params.status);
    if (params.source) search.set("source", params.source);
    if (params.q && params.q.trim()) search.set("q", params.q.trim());

    const res = await api.get<DemandPage>(`/demands?${search.toString()}`);
    return res.data;
  },

  /** ADMIN or owner: fetch one demand */
  getDemand: async (id: number): Promise<Demand> => {
    const res = await api.get<Demand>(`/demands/${id}`);
    return res.data;
  },

  /** ADMIN: update demand status */
  updateStatus: async (
    id: number,
    status: DemandStatus,
    adminNote?: string,
    rejectionReason?: string
  ): Promise<Demand> => {
    const res = await api.patch<Demand>(`/demands/${id}/status`, {
      status,
      adminNote,
      rejectionReason,
    });
    return res.data;
  },

  /** ADMIN: assign to a specific admin, or to the caller when adminId is omitted */
  assign: async (id: number, adminId?: string): Promise<Demand> => {
    const res = await api.patch<Demand>(`/demands/${id}/assign`, adminId ? { adminId } : {});
    return res.data;
  },

  /** ADMIN: change priority */
  updatePriority: async (id: number, priority: DemandPriority): Promise<Demand> => {
    const res = await api.patch<Demand>(`/demands/${id}/priority`, { priority });
    return res.data;
  },

  /** ADMIN: convert demand to project (existing client) */
  convertToProject: async (
    id: number,
    clientId: number,
    coordinates?: { latitude?: number; longitude?: number }
  ): Promise<{ id: number }> => {
    const res = await api.post<{ id: number }>(`/demands/${id}/convert-to-project`, {
      clientId,
      latitude: coordinates?.latitude,
      longitude: coordinates?.longitude,
    });
    return res.data;
  },

  /** ADMIN: promote a PUBLIC demand into a client + project in one shot */
  promotePublicAndConvert: async (
    id: number,
    coordinates?: { latitude?: number; longitude?: number }
  ): Promise<{ id: number }> => {
    const res = await api.post<{ id: number }>(
      `/demands/${id}/promote-public-to-client`,
      {
        latitude: coordinates?.latitude,
        longitude: coordinates?.longitude,
      }
    );
    return res.data;
  },

  /** ADMIN: send client invitation by email and/or SMS */
  sendInvitation: async (
    id: number,
    clientEmail: string,
    clientName: string,
    projectId: number | null,
    phoneNumber?: string,
    message?: string,
    sendEmail: boolean = true,
    sendSms: boolean = true
  ): Promise<{ success: string; message: string }> => {
    const res = await api.post<{ success: string; message: string }>(`/demands/${id}/send-invitation`, {
      demandId: id,
      projectId,
      clientEmail,
      clientName,
      phoneNumber,
      message,
      sendEmail,
      sendSms,
    });
    return res.data;
  },
};

export default demandService;
